src/ir/transform.cc

```diff
@@ -164,34 +167,57 @@ void PassContext::RegisterConfigOption(const char* key, uint32_t value_type_inde
 
 PassContext PassContext::Create() { return PassContext(make_object<PassContextNode>()); }
 
-void PassContext::InstrumentSetUp() const {
+void PassContext::InstrumentEnterPassContext() {
   auto pass_ctx_node = this->operator->();
   if (pass_ctx_node->instruments.defined()) {
-    for (instrument::PassInstrument pi : pass_ctx_node->instruments) {
-      pi->SetUp();
+    try {
+      for (instrument::PassInstrument pi : pass_ctx_node->instruments) {
+        pi->EnterPassContext();
+      }
+    } catch (const Error& e) {
+      LOG(INFO) << "Pass instrumentation entering pass context failed.";
+      LOG(INFO) << "Disable pass instrumentation.";
+      pass_ctx_node->instruments.clear();
+      throw e;
     }
   }
 }
 
-void PassContext::InstrumentTearDown() const {
+void PassContext::InstrumentExitPassContext() {
   auto pass_ctx_node = this->operator->();
   if (pass_ctx_node->instruments.defined()) {
-    for (instrument::PassInstrument pi : pass_ctx_node->instruments) {
-      pi->TearDown();
+    try {
+      for (instrument::PassInstrument pi : pass_ctx_node->instruments) {
+        pi->ExitPassContext();
+      }
+    } catch (const Error& e) {
+      LOG(INFO) << "Pass instrumentation exiting pass context failed.";
+      pass_ctx_node->instruments.clear();
+      throw e;
     }
   }
 }
 
 bool PassContext::InstrumentBeforePass(const IRModule& ir_module, const PassInfo& pass_info) const {
   auto pass_ctx_node = this->operator->();
-  if (pass_ctx_node->instruments.defined()) {
+  if (!pass_ctx_node->instruments.defined()) {
+    return true;
+  }
+
+  const bool pass_required = PassArrayContains(pass_ctx_node->required_pass, pass_info->name);
+  bool should_run = true;
+  if (!pass_required) {
+    const Array<instrument::PassInstrument>& instruments = pass_ctx_node->instruments;
+    should_run &= std::all_of(instruments.begin(), instruments.end(),
```

Comment:
this is a bit counter to the stl, but I actually prefer if code like this is just explicitly written out. then it's clear what happens if any of `pi->ShouldRun` returns early.

Here, it might be slightly better to explicitly invoke all ShouldRun(), even if one returns False. Then, passes that simply implement some type of logging or timing can still include entries for those passes that didn't run. i'm happy to be swayed either way on this, but it seems more useful to me that way.

Reply:
Removed all_of (much like llvm's code above now).

For the second question, personally, I prefer to be consistent with llvm's API and its logics like above.
If there are strong desire to do that in the future, I will change it.

Reply:
thanks! I think reading the LLVM API again, it actually does what I want, so no issue.